server/jamovi/server/dataset/duckdataset.py

```diff
@@ -172,26 +181,58 @@ def __iter__(self) -> Iterator[DuckColumn]:
         for column in self._columns_by_index:
             yield column
 
-    def set_value(self, row: int, column: int, value, initing=False):
+    def set_value(self, row: int, column_index: int, value, initing=False):
         """set a value in the data set"""
-        col = self._columns_by_index[column]
-
-        if isinstance(value, str) and col.data_type is DataType.TEXT and col.has_levels:
+        # Question: Did you want the signature to be changed to take row and column as a tuple?
+        col = self._columns_by_index[column_index]
+        value = self._convert_value_to_raw(value, col)
+        self._write_buffer.stage(CellCoordinate(row_index=row, col_iid=col.iid), value=value)
+
+    def set_values(self, values:dict[CellCoordinate, CellValue]):
+        self._set_values(values, index_name="index")
+
+    def _set_values(self, values:dict[CellCoordinate, CellValue], index_name):
+        
+        # Values written one column at a time base on the assumption that cells 
+        # to be set are not neccessarily adjacent.
+         
+        # Note: Could be improved by updating groups of columns that have 
+        # updates for all the same rows as each other
```

Comment:
yeah, there's lots of interesting optimisations we could do. for the present implementation, as the number of columns in the dataset approaches the size of the cache, the speed up approaches zero.

it will be interesting to profile different scenarios with different cache sizes.